packages/playwright-core/src/server/injected/roleUtils.ts

```diff
@@ -461,6 +461,60 @@ export function getElementAccessibleDescription(element: Element, includeHidden:
   return accessibleDescription;
 }
 
+// https://html.spec.whatwg.org/multipage/input.html#the-input-element
+export const kInputValidityAttributes = ['required', 'pattern', 'min', 'max', 'minlength', 'maxlength', 'step', 'type', 'value'];
+
+export function hasValidationAttributes(element: Element): boolean {
+  if (!(element instanceof HTMLInputElement))
+    return false;
+
+  return kInputValidityAttributes.some(attr => element.hasAttribute(attr));
+}
+
+// https://www.w3.org/TR/wai-aria-1.2/#aria-invalid
+export const kAriaInvalidRoles = ['application', 'checkbox', 'combobox', 'gridcell', 'listbox', 'radiogroup', 'slider', 'spinbutton', 'textbox', 'tree', 'columnheader', 'rowheader', 'searchbox', 'switch', 'treegrid'];
+
+export function getAriaInvalid(element: Element): 'false' | 'true' | 'grammar' | 'spelling' {
+  const role = getAriaRole(element) || '';
+  if (!role || !kAriaInvalidRoles.includes(role))
+    return 'false';
+  if (element instanceof HTMLInputElement && hasValidationAttributes(element) && element.validity)
```

Comment:
Perhaps check both `aria-invalid` and `element.validity?.valid === false`, and consider the element invalid when either of the indicators is present?

Reply:
1. I understood it as being asked to remove hasValidationAttributes and evaluate only with `element.validity?.valid === false` without assessing the `kAriaInvalidRoles attributes`. Would that be correct? :)

2. According to the [spec](https://www.w3.org/TR/wai-aria-1.2/#aria-invalid), it seems that the return values of the `getAriaInvalid` function should be `'false' | 'true' | 'grammar' | 'spelling'`.  
Rather than evaluating `validity` within this function, I think it would be better to separate it into a dedicated function, such as `getValidityInvalid`. I look forward to your feedback on this! :)